metrics-datadog/src/main/scala/com/ovoenergy/effect/Datadog.scala

```diff
@@ -3,42 +3,90 @@ package com.ovoenergy.effect
 import java.net.InetSocketAddress
 
 import cats.effect._
+import com.ovoenergy.effect.Events.Event
 import com.ovoenergy.effect.Metrics.Metric
 import fs2.Chunk.array
 import fs2.io.udp._
 
 object Datadog {
 
+  // just for readability
+  type UTF8Bytes = Array[Byte]
+
+  private implicit class StringOps(s: String) {
+    def utf8Bytes: UTF8Bytes = s.getBytes("UTF-8")
+  }
+
+  private implicit class ByteArrayOps(bs: UTF8Bytes) {
+    def sep(char: Char)(other: UTF8Bytes): UTF8Bytes = (bs :+ char.toByte) ++ other
+    def colon: UTF8Bytes => UTF8Bytes = sep(':')
+    def pipe: UTF8Bytes => UTF8Bytes = sep('|')
+  }
+
   case class Config(
     agentHost: InetSocketAddress,
     metricPrefix: Option[String],
     globalTags: Map[String, String]
   )
 
+  /**
+   * Apparently the maximum UDP packet size is 65535 bytes (at the absolute maximum)
+   * and we have many different strings in a packet so we only allow each one to be 1Kb
+   */
+  val maxStringLength = 1000
+
+  /**
+   * A basic sanity check for the maximum number of tags to send to Datadog
+   * 100 x 2000 (key + value size) = 20Kb so well within a reasonable budget
+   */
+  val maximumTagCount = 100
+
   /**
    * Datadog enforces all metrics must start with a letter
    * and not contain any chars other than letters, numbers and underscores
    */
-  private[effect] def filterName(s: String): String =
-    s.dropWhile(!_.isLetter).replaceAll("[^A-Za-z0-9\\.]+", "_")
+  private[effect] def filterName(s: String): UTF8Bytes =
+    s.dropWhile(!_.isLetter).replaceAll("[^A-Za-z0-9.]+", "_").take(maxStringLength).utf8Bytes
 
   /**
    * More lenient filtering for tag values,
    * we allow alpha numeric characters, slashes, hyphens and numbers
    */
-  private[effect] def filterValue(s: String): String =
-    s.replaceAll("[^A-Za-z0-9\\./\\-]+", "_")
+  private[effect] def filterValue(s: String): UTF8Bytes =
+    s.replaceAll("[^A-Za-z0-9./\\-]+", "_").take(maxStringLength).utf8Bytes
+
+  /**
+   * Datadog receives events as binary byte arrays then converts them into UTF-8 strings
+   * https://github.com/DataDog/datadog-agent/blob/21a80ab80de389adb9c74e3e9a7162f83fda3e0c/pkg/dogstatsd/parse_events.go#L62
+   * As such we obtain the byte values from the string in UTF-8
+   */
+  private[effect] def filterEventText(s: String): UTF8Bytes =
+    s.take(maxStringLength).toCharArray.flatMap {
```

Comment:
Is this the same as `s.take(maxStringLength).replaceAllLiterally("\r", "\n").utf8Bytes`?

Reply:
it also does `\n` => `\\n` but yes I should absolutely just make this a replaceAll :+1:  Originally I was doing a bunch of other filtering ops here